src/database.ts

```diff
@@ -3241,41 +3275,42 @@ class Database extends common.GrpcServiceObject {
           transaction!.excludeTxnFromChangeStreams();
         }
 
-        const release = () => {
+        // Our span should only be ended if the
+        // transaction either errored or was ended.
+        transaction!.once('error', err => {
+          setSpanError(span, err!);
           span.end();
+        });
+
+        transaction!.once('end', err => {
+          setSpanError(span, err!);
+          span.end();
+        });
+
+        const release = () => {
           this.pool_.release(session!);
         };
 
         const runner = new TransactionRunner(
           session!,
           transaction!,
-          (err, resp) => {
-            if (err) {
-              setSpanError(span, err!);
-            }
-            span.end();
-            runFn!(err, resp);
-          },
+          runFn!,
           options
         );
 
         runner.run().then(release, err => {
-          if (err) {
-            setSpanError(span, err!);
-          }
+          setSpanError(span, err);
```

Comment:
Can't we add 
```
runner.run()
.then( -------->)
.finally(() => { span.end(); });
```

Reply:
The problem with that, if the code has that retry down below of which on the retry, we have to wait until the customer is done using the transaction hence the `.once('end', () => span.end)` otherwise runTransaction's spans are going to end prematurely before completion

Reply:
@surbhigarg92 the other problem with using runner.run().finally is that the final span will not be ever generated until after database.runTransaction() has been exited, not when transaction.end() or .commit() have been invoked. Yours and @alkatrivedi's visual confirmation tests won't work until you wait. The suggested change causes incorrect behavior because now even after the customer has correctly used transaction.end() or transaction.commit() runTransaction's span won't be ended until their callback exists :-(